Upgrade dependencies and migrate to pnpm

bryanjtc opened the pull request:
Fixes #1084

bryanjtc commented:
Almost all peer dependencies were modified. Some were relaxed a bit and others upgraded to the next major version to match the minimum required of some plugins and configs.

bryanjtc commented:
I also checked if the peer dependencies were the minium required to support node 18.

bryanjtc commented:
@sobolevn Can you check this?

sobolevn reviewed (commented):
Thanks a lot! Great work!

What do you think about dropping `yarn` in favor of some other tool?
`pnpm` / `npm` / whatever else?

I don't think that `yarn` is the way to go in 2022 (almost 2023).

bryanjtc commented:
Yarn introduced PnP in v2. It removes node_modules in favor of a cache folder to store the dependencies in zips. In both yarn pnp and pnpm all peer dependenciea have to be declared explicity in the package.json. I can try to migrate the repo to yarn pnp.

bryanjtc commented:
https://yarnpkg.com/features/pnp

bryanjtc commented:
I prefer yarn pnp over pnpm, npm because it's the fastest alternative.

bryanjtc commented:
![image](https://user-images.githubusercontent.com/49354825/207132729-88d71efa-a16e-4647-823c-d24bc3426ae2.png)

bryanjtc commented:
@sobolevn I tried enabling pnp for yarn, but there were some issues. So I migrated to pnpm.

sobolevn commented on pnpm-workspace.yaml:
```diff
@@ -0,0 +1,2 @@
+packages:
+  - 'packages/**'
```
Shouldn't `packages` allow to have the same dev dependencies everywhere? 🤔

bryanjtc replied:
pnpm doesn’t use dependency hoisting:

When installing dependencies with npm or Yarn Classic (v1), all packages are hoisted to the root of the modules directory. As  a result, source code has access to dependencies that are not added as dependencies to the project.
By default, [pnpm](https://pnpm.io/motivation#creating-a-non-flat-node_modules-directory) uses symlinks to add only the direct dependencies of the project into the root of the modules directory.

In practice it means that if you have a package A that imports a package B (import something from 'B') but doesn’t explicitly specify B in the dependencies or devDependencies, then the execution will fail.

sobolevn reviewed (commented):
Thanks a lot! Looks great!

sobolevn reviewed (approved):
Let's give it a try!

I will merge and release the new version tomorrow.
Thank you!

bryanjtc commented:
@sobolevn Dependabot doesn't support pnpm, so I replaced it with renovate. You have to install the github app on the repo. https://github.com/apps/renovate
Renovate works better with pinned dependencies, so I pin them.
Fixed a peer dependency warning.

sobolevn commented:
Thank you for your amazing work! Sorry, it took me a while to get back to it :)
